[quant][pt2] Preserve source_fn_stack after QAT fusion

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110899

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

andrewor14 committed:
[quant][pt2] Preserve source_fn_stack after QAT fusion

Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -528,6 +528,58 @@ def forward(self, x):
+
+        # Assert these nodes have the right source partitions
+        node_name_to_source_partition = {
+            "_param_constant0": "l__self___backbone_conv",
```
should we match the type string instead? that feels more stable than the name

jerryzh168 commented on test/quantization/pt2e/test_quantize_pt2e_qat.py:
```diff
@@ -528,6 +528,58 @@ def forward(self, x):
+        # Assert these nodes have the right source partitions
+        node_name_to_source_partition = {
+            "_param_constant0": "l__self___backbone_conv",
+            "_param_constant1": "l__self___backbone_conv",
```
also node_names are free to change, so we should not rely on it I think

andrewor14 replied:
sounds good, let me see what I can do

jerryzh168 commented on torch/ao/quantization/pt2e/qat_utils.py:
```diff
@@ -571,8 +574,13 @@ def _fuse_conv_bn_qat_helper(m: GraphModule, is_cuda: bool) -> GraphModule:
-        (replacement_conv_node, replacement_bn_node, replacement_getitem_node) =\
-            _get_conv_bn_getitem_nodes(r.replacements)
+        # TODO: rewrite this to return a map
+        (
+            replacement_conv_node,
+            replacement_bn_node,
+            replacement_getitem_node,
+            replacement_relu_node,
+        ) = _get_conv_bn_getitem_nodes(r.replacements)
```
I guess this is for identifying the nodes, I feel https://github.com/pytorch/pytorch/pull/110743 should help

jerryzh168 reviewed (approved):
maybe we can extend https://github.com/pytorch/pytorch/pull/110743 to work with subgraph rewriter and query node by name

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

[ghstack-poisoned]

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

andrewor14 committed:
Update on "[quant][pt2] Preserve source_fn_stack after QAT fusion"


Test Plan:
python test/test_quantization.py TestQuantizePT2EQAT.test_qat_preserve_source_fn_stack

Reviewers: jerryzh168, kimishpatel

Subscribers: jerryzh168, kimishpatel, supriyar

Differential Revision: [D50101253](https://our.internmc.facebook.com/intern/diff/D50101253)

[ghstack-poisoned]

andrewor14 commented:
@andrewor14 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50101253).

andrewor14 commented:
@pytorchbot merge

kimishpatel commented on torch/ao/quantization/pt2e/qat_utils.py:
```diff
@@ -359,14 +363,14 @@ def _get_fused_convbn_q_dq_nodes(nodes: List[Node]) -> Tuple[Node, Node]:
     return (q_node, dq_node)
 
-def _get_conv_bn_getitem_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node]:
+def _get_conv_bn_getitem_nodes(nodes: List[Node]) -> Tuple[Node, Node, Node, Optional[Node]]:
```
should also be renamed to _get_conv_bn_getitem_relu_nodes

kimishpatel commented on torch/ao/quantization/pt2e/qat_utils.py:
```diff
@@ -694,9 +710,29 @@ def _fold_conv_bn_qat_helper(m: GraphModule, is_cuda: bool) -> GraphModule:
+            replacement_relu_node,
+        ) = _get_conv_bn_getitem_nodes(r.replacements)
+
+        for original_node in _filter_nodes_map(r.nodes_map).values():
```
Andrew, why do you need to copy the metadata?

andrewor14 replied:
This is for lowering models trained by QAT. We found that the `source_fn` was not preserved and all convs in the model ended up being in the same partition `conv2d`, but they should be in different partitions
